File: java/pdfclown.samples.cli/src/org/pdfclown/samples/cli/NamesParsingSample.java
package org.pdfclown.samples.cli;

import java.util.Map;

import org.pdfclown.documents.Document;
import org.pdfclown.documents.NamedDestinations;
import org.pdfclown.documents.Names;
import org.pdfclown.documents.Page;
import org.pdfclown.documents.interaction.navigation.document.Destination;
import org.pdfclown.files.File;
import org.pdfclown.objects.PdfReference;

/**
  This sample demonstrates <b>how to inspect the object names</b> within a PDF document.

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @since 0.0.8
  @version 0.1.2, 02/04/12
*/
public class NamesParsingSample extends Sample
{
  @Override
  public boolean run()
  {
    // 1. Opening the PDF file...
    File file;
    {
      String filePath = promptFileChoice("Please select a PDF file");
      try
      {file = new File(filePath);}
      catch(Exception e)
      {throw new RuntimeException(filePath + " file access error.",e);}
    }
    Document document = file.getDocument();

    // 2. Named objects extraction.
    Names names = document.getNames();
    if(names == null)
    {System.out.println("\nNo names dictionary.");}
    else
    {
      System.out.println("\nNames dictionary found (" + names.getContainer().getReference() + ")");

      NamedDestinations namedDestinations = names.getDestinations();
      if(namedDestinations == null)
      {System.out.println("\nNo named destinations.");}
      else
      {
        System.out.println("\nNamed destinations found (" + namedDestinations.getContainer().getReference() + ")");

        // Parsing the named destinations...
        for(Map.Entry<String,Destination> namedDestination : namedDestinations.entrySet())
        {
          String key = namedDestination.getKey();
          Destination value = namedDestination.getValue();

          System.out.println("  Destination '" + key + "' (" + value.getContainer().getReference() + ")");

          System.out.print("    Target Page: number = ");
          Object pageRef = value.getPage();
          if(pageRef instanceof Integer) // NOTE: numeric page refs are typical of remote destinations.
          {System.out.println(((Integer)pageRef) + 1);}
          else // NOTE: explicit page refs are typical of local destinations.
          {
            Page page = (Page)pageRef;
            System.out.println((page.getIndex() + 1) + "; ID = " + ((PdfReference)page.getBaseObject()).getId());
          }
        }

        System.out.println("Named destinations count = " + namedDestinations.size());
      }
    }

    return true;
  }
}

File: dotNET/pdfclown.samples.cli/src/org/pdfclown/samples/cli/NamesParsingSample.cs
using org.pdfclown.documents;
using org.pdfclown.documents.interaction.navigation.document;
using org.pdfclown.files;
using org.pdfclown.objects;

using System;
using System.Collections.Generic;

namespace org.pdfclown.samples.cli
{
  /**
    <summary>This sample demonstrates how to inspect the object names within a PDF document.</summary>
  */
  public class NamesParsingSample
    : Sample
  {
    public override bool Run(
      )
    {
      // 1. Opening the PDF file...
      string filePath = PromptPdfFileChoice("Please select a PDF file");
      File file = new File(filePath);
      Document document = file.Document;

      // 2. Named objects extraction.
      Names names = document.Names;
      if(names == null)
      {Console.WriteLine("No names dictionary.");}
      else
      {
        Console.WriteLine("Names dictionary found (" + names.Container.Reference + ")");

        NamedDestinations namedDestinations = names.Destinations;
        if(namedDestinations == null)
        {Console.WriteLine("No named destinations.");}
        else
        {
          Console.WriteLine("Named destinations found (" + namedDestinations.Container.Reference + ")");

          // Parsing the named destinations...
          foreach(KeyValuePair<PdfString,Destination> namedDestination in namedDestinations)
          {
            PdfString key = namedDestination.Key;
            Destination value = namedDestination.Value;

            Console.WriteLine("Destination '" + key.Value + "' (" + value.Container.Reference + ")");

            Console.Write("    Target Page: number = ");
            object pageRef = value.PageRef;
            if(pageRef is Int32) // NOTE: numeric page refs are typical of remote destinations.
            {Console.WriteLine(((int)pageRef) + 1);}
            else // NOTE: explicit page refs are typical of local destinations.
            {
              Page page = (Page)pageRef;
              Console.WriteLine((page.Index + 1) + "; ID = " + ((PdfReference)page.BaseObject).Id);
            }
          }

          Console.WriteLine("Named destinations count = " + namedDestinations.Count);
        }
      }

      return true;
    }
  }
}